Patch-Source: https://github.com/davatorium/rofi/pull/1570
From 2832bdc706c444d0261975559a8ad4174a81e559 Mon Sep 17 00:00:00 2001
From: Jakub Jirutka <jakub@jirutka.cz>
Date: Sat, 22 Jan 2022 17:52:28 +0100
Subject: [PATCH] Add -combi-display-format

Implement a new option -combi-display-format (analogous to
-drun-display-format) that allows to change position appearance of
the mode name in the combi entries.
---
 config/config.c        |  1 +
 doc/rofi.1.markdown    | 14 ++++++++++++++
 doc/test_xr.txt        |  2 ++
 include/settings.h     |  2 ++
 source/dialogs/combi.c | 19 +++++++++++++++++--
 source/xrmoptions.c    |  6 ++++++
 6 files changed, 42 insertions(+), 2 deletions(-)

diff --git a/config/config.c b/config/config.c
index d724a7cc..bae72776 100644
--- a/config/config.c
+++ b/config/config.c
@@ -133,6 +133,7 @@ Settings config = {
     .plugin_path               = PLUGIN_PATH,
     .max_history_size          = 25,
     .combi_hide_mode_prefix    = FALSE,
+    .combi_display_format      = "{mode} {text}",

     .matching_negate_char      = '-',

diff --git a/doc/rofi.1.markdown b/doc/rofi.1.markdown
index d70ea4ab..bfb8c673 100644
--- a/doc/rofi.1.markdown
+++ b/doc/rofi.1.markdown
@@ -544,6 +544,20 @@ To get one merge view, of `window`,`run`, and `ssh`:
 **NOTE**: The i3 window manager dislikes commas in the command when specifying an exec command.
 For that case, `#` can be used as a separator.

+`-combi-display-format`
+
+The format string for entries in the `combi` dialog:
+
+* **mode**: the mode display name
+* **text**: the entry text
+
+Pango markup can be used to formatting the output.
+
+    Default: {mode} {text}
+
+Note: This setting is ignored if `combi-hide-mode-prefix` is eanbled.
+
+
 ### History and Sorting

 `-disable-history`
diff --git a/include/settings.h b/include/settings.h
index a5dcf5fc..e6e21af3 100644
--- a/include/settings.h
+++ b/include/settings.h
@@ -154,6 +154,8 @@ typedef struct {
   /** Maximum history length per mode. */
   unsigned int max_history_size;
   gboolean combi_hide_mode_prefix;
+  /** Combi format display */
+  char *combi_display_format;

   char matching_negate_char;

diff --git a/source/dialogs/combi.c b/source/dialogs/combi.c
index 9f66c7a6..ebd50fe6 100644
--- a/source/dialogs/combi.c
+++ b/source/dialogs/combi.c
@@ -35,6 +35,7 @@
 #include <stdlib.h>

 #include "mode-private.h"
+#include "widgets/textbox.h"
 #include <dialogs/dialogs.h>
 #include <pango/pango.h>
 #include <theme.h>
@@ -221,12 +222,26 @@ static char *combi_mgrv(const Mode *sw, unsigned int selected_line, int *state,
     if (selected_line >= pd->starts[i] &&
         selected_line < (pd->starts[i] + pd->lengths[i])) {
       char *retv;
+      int entry_state = 0;
       char *str = retv = mode_get_display_value(pd->switchers[i].mode,
                                                 selected_line - pd->starts[i],
-                                                state, attr_list, TRUE);
+                                                &entry_state, attr_list, TRUE);
       const char *dname = mode_get_display_name(pd->switchers[i].mode);
       if (!config.combi_hide_mode_prefix) {
-        retv = g_strdup_printf("%s %s", dname, str);
+
+        if (!(entry_state & MARKUP)) {
+          char *tmp = str;
+          str = g_markup_escape_text(tmp, -1);
+          g_free(tmp);
+          entry_state |= MARKUP;
+        }
+        *state |= entry_state;
+
+        retv = helper_string_replace_if_exists(
+            config.combi_display_format,
+            "{mode}", dname,
+            "{text}", str,
+            NULL);
         g_free(str);

         if (attr_list != NULL) {
diff --git a/source/xrmoptions.c b/source/xrmoptions.c
index bb04fb53..ccd0aa04 100644
--- a/source/xrmoptions.c
+++ b/source/xrmoptions.c
@@ -356,6 +356,12 @@ static XrmOption xrmOptions[] = {
      NULL,
      "Hide the prefix mode prefix on the combi view.",
      CONFIG_DEFAULT},
+    {xrm_String,
+     "combi-display-format",
+     {.str = &config.combi_display_format},
+     NULL,
+     "Combi format string. (Supports: mode, text)",
+     CONFIG_DEFAULT},
     {xrm_Char,
      "matching-negate-char",
      {.charc = &config.matching_negate_char},
--
2.16.4

